### PR TITLE
Add additional device features

### PR DESCRIPTION
**Pull request by neilharvey94044**
I've made changes to complete the TODOs you left and add methods to support additional features in SDRplay devices, with a focus on RSP1A but enabling additional capabilities and expansion for all SDRplay devices.  The changes I've made here are dependencies to a larger set of changes I've made for sdrtrunk to do the following:

1. Sample Size and IF Bandwidth are aligned, same as done in SDRuno from SDRplay.

2. Defaults to using IF Zero for all Sample Sizes.

3. Decimation has been implemented with allowable values based upon Sample Size.

4. Automatic Gain Control (AGC) has been implemented.

5. Defaults to DC Offset calibration enabled.

6. For RSP1A the Bias-T, FM Broadcast Notch, and Digital Audio Broadcast Notch options have been implemented.

7.  For RSP1A, the number of LNA RF Gain Reduction states matches the RSP1A capabilities.

8.  Additional SDRplay device capabilities can be more easily added in future releases.

These changes should allow users to get up and running with SDRplay devices in sdrtrunk with fewer challenges.  I modeled the settings after SDRPlay's SDRuno software.

Feel free to email me directly, my email address is published on Github.

### CODE REVIEW

**Comment by samphonic** [comment]
Wow, awesome!

I've been super distracted with life-stuff since before the holidays so I haven't had a chance to dig back into any of this.  I don't know if I will have a chance soon to test the changes myself (one of the life-things was moving and we're still unpacking), but I'll give the code a look over soon.  

You didn't mention, but is it safe to assume you successfully tested your changes on an RSP1A?

(I'll need to refamiliarize myself with some of the code since it's been a while, but I think if I merge this it should get pushed out to the Maven repos automatically, though I just got some email from JFrog about how that might have changed so I'll have to look at that too).

**Comment by neilharvey94044** [comment]
Yes, I've tested it with the RSP1A.  Works great.  I've been testing as I develop.  I've also created a PR for sdrtrunk that leverages the changes here.
Screenshot 2 sdrplay

**Comment by samphonic** [comment]
Looked over the changes (sorry it took a while), and everything looks great.  On my todo list (eventually) is to add RSPduo support, so your changes will make that a much quicker process.  I'll merge this now and try to make sure it's all published to the various repos.  If there's anything you need from me to make the sdrplay PR go smoothly let me know.

**Comment by neilharvey94044** [comment]
Not related to the pull request...did you ever figure out how the SDRplay debug api call actually provides debug information?  I've been looking for log files it may have produced, but don't seem to find any.   Also, no obvious connection to a callback or other mechanism for obtaining the logs is described in their documentation.

I went out and bought an RSPdx and am finding that the AGC selection throws an exception - sdrplay_api_OutOfRange.  This doesn't happen with the RSP1A.  Hoping the debug function can help me find the problem.

**Comment by samphonic** [comment]
I won't be able to check on my machine to get you details until tonight,
but you have to use an external program to view the debug output.  I
*believe* it was WinDbg.

So you'd start up WinDbg and then enable the debug output in the API.

Let me know if that helps, if not I'll try to go through the process
tonight so I can remember exactly how it works.

**Comment by neilharvey94044** [comment]
That would be great!  I've installed WinDbg, but the path forward does not look obvious.  This must be a Windows logging/debugging API I've not seen before, designed for hardware debugging.  I tried attaching to the SDRPlay API service, but can't find the PID for it.  I assume it's the API I need to attach to, or an API generated log I need to load.

**Comment by samphonic** [comment]
Whoops, sorry about that-- I wasn't at my computer so I couldn't check what I'd used and Google wasn't helping.  The correct application is DebugView ([available here](https://docs.microsoft.com/en-us/sysinternals/downloads/debugview)).  Just open that up (make sure you use the 32 or 64 bit one depending on system/service), and when you call `DebugEnable` you should start to see output from the SDRplay service in DebugView.

There are some options on top for filtering and turning off autoscroll that I found pretty useful.  There's not always a ton of helpful info there, but hopefully it'll help you with what you're encountering.

**Comment by neilharvey94044** [comment]
Thanks!

**Comment by neilharvey94044** [comment]
I'm running into a real challenge with the jnr library.  The memory mapping/alignment doesn't seem to be accurate deeper in the parameter structure.  I've been studying the structs both in c and Java/jnr looking for misalignment or incorrect sizing.  Not finding anything after searching in detail several times.  Really frustrating.  I finally decided to try tailoring the sample c program to verify what specific API calls output in the debug information - for example setting the rfNotch.  Everytime I ran it in Java using your short JNRMain.java it would report no errors, but in the debug would show a zero after the call, each and every time, even when I was trying to enable the notch.  When running the c program...different behavior, it's actually working and I see the "1" or "0" depending on how I've set it.  This confirms the memory alignment is not working in Java/jnr.

Long story short, do you have any tricks for understanding the memory map in jnr?  I noticed the Struct class has a toString() but it throws an exception.  Probably an indicator of the problem, but I don't seem to see what it is, each Struct is being handled the same way and aligns with the c header files perfectly.  I've mostly just followed your patterns using jnr, not developed any real understanding of how it's working under the covers.

I tried running Eclipse debug, but it just hangs.  Actually now I recall according to the API disabling heartbeats might solve this hanging.  My next step.

Any suggestions appreciated.

**Comment by samphonic** [comment]
Yeah, the wrapper libraries have been a pain.  I tried SWIG, JNA, and JNI and they all failed in weird subtle ways before I eventually got JNR to work despite the relative lack of documentation.

In the intervening 4 months, I've forgotten some of the troubleshooting steps I took, but some thoughts in case they help a little:

- I don't fully understand it myself, but I think at a high level JNR works by:
  - Providing some custom types (e.g. `Unsigned8`) so that field values are stored in-memory in a way that C can read them.
  - Injecting itself in/around the JVM's memory management so that fields are stored at memory addresses that can be passed to C function calls.
  - Doing some kind of magic when you call a C function so that pointers are passed to the memory addresses it's been handling.

- Initialization of the Struct is very sensitive.  As far as I can tell/remember you seem to have done everything the same as I did for `Rsp1aParamsT`, but things like initializing fields inside or outside the constructor made a difference.  Nested Structs need the `inner(struct)` format, but that's there.  (Weirdly I was able to successfully use some Structs without even defining all of the fields, so the Struct doesn't necessarily need to be complete to work (no idea what happens with the undefined fields).)
- I was able to debug in NetBeans I think, but all I really was able to glean was the memory address of certain fields.  If the address _changes_ when you set/get a value from that field (or get/set the struct it's part of maybe?), it's _not_ properly pointing to a real C memory address (e.g. the JVM is managing it without the JNR mapping magic).
- Related to this is some of the weird [nonsense I wrote here](https://github.com/Sammy1Am/SDRplayJava/blob/17347cdc78c7c128f993528e9b21787e8a72dcb0/src/main/java/SDRplayJava/JNRMain.java#L86-L95).  The JNR examples seem to imply you could call `sdrplay_api_GetDeviceParams` and pass a `DeviceParamsT` struct to be filled with the correct return values; but for whatever reason I had to go through this `PointerByReference` + `locallyConstructedStruct.useMemory()` route to get the field memory-addresses to be properly allocated.  Once again though, it looks like you followed the same patterns I was using before so I don't know what could be amiss.

You can also use the SDRplay header files installed with the API (in your Program Files folder), and a C debugger (like Visual Studio) to debug the API itself.  You can set a breakpoint somewhere in the API and use that to poke around to see if the DeviceParamsT Struct has all the values you expect and maybe the memory assignments you see in the Java debug.  I was able to do this at one point, but I'm not sure it ended up actually helping me with anything 😝.

It's most likely a case of some fields in Java not properly getting the memory-address magic assigned to them so that when C receives your new params that field doesn't have (any?) value.  Hopefully this has been a little helpful.  Keep me posted on your progress and I'll let you know if I can think of anything else.  Good luck!

**Comment by neilharvey94044** [comment]
I really appreciate your taking the time to write this.

Yesterday I was able to stop the heartbeat in the API and that allowed debugging in Eclipse.  Still not a resolution, but at least now I can see how the Structs are organized and what's in each field in Java.  It appears to keep information about each field, offset into the immediate struct, native type, size in bytes, etc..  I'm guessing at the time a native call occurs it maps the values of those fields into the native compatible memory.  What I need to figure out next is if the mapping of data is the problem, the timing of mapping, or the organization of the data in Java.  As you know even one byte off will throw the remainder of the struct off.  As yet, I haven't seen the actual values in the java jnr objects, I think it just uses the object to figure out where to put the values in native memory.

I've also been in contact with SDRPlay and they've shared some insight into API constraints.  Not a lot, but some.  They've also acknowledged their API specification is lacking these constraints and are considering updating it.  That will take some time though, assuming they decide to do it.

Until I've sorted this out I have cancelled my pull request on sdrtrunk.

Thanks!

**Comment by samphonic** [comment]
I may be confusing JNR with one of the other Java native wrappers, but I'm fairly confident that in at least some cases the Java object values are stored in the native-accessible memory even before a native call is made (which is where that `useMemory()` call comes in).  You should be able to see the values in the Java debugger (for example the DeviceT objects when the system is initializing and listing out available devices).  That said, I think there may also be situations where the object in Java is created first and then its fields pointed to the native memory later, but from what I remember I think those were situations where I was having issues before. 🤷‍♂️

It's ultimately up to DSheirer (and up to you since it's your pull request), but I personally think there would be some value to the sdrtrunk community to have SDRplay support integrated even if not all of the features are complete.  Based on the comments on my PR, it seems like people were able to get it up and running decently even with very rudimentary support (I know as an end user I'd rather have basic support for my radio than no support at all).  That is to say, if you hit too much of a wall with some of the details, it's probably still worth going through with the PR.

But hopefully, of course, you'll have everything figured out soon and working 😉.  I wish I could help debug more directly, but it's still going to be a bit before I have a radio available at home again.

**Comment by neilharvey94044** [comment]
I hear you and agree, my RSP1A and RSPDX both are working in sdrtrunk with this code for P25.  But before I put it back out there I need to get some level of confidence of what is working by design and what is just working by chance.   Now that I have the SDRPlay debugging mechanism working I can see what the RSP is actually doing.  The JNR foundation needs to be solid so other functionality can be added.  I closed the pull request because I don't want to take Denny's time while I'm working these issues out.

You did an amazing job of getting the JNR working for such a complex set of structs and calls with no documentation.  Getting anything working between a managed environment and a native environment is a challenge, but doing it with no documentation of the framework is just amazing.  I just need to keep at it and work out the wrinkles.  I am a mostly-retired old guy so have time to put into it.  I'm also having fun with it, learning a lot and thankful that you are providing insight.  😀

With regard to JNR behavior, I think what it's doing is using the Struct objects as meta-data and using the native memory as the "backing store".  So the actual data being passed by get() and set() calls is being put and taken from native memory and never actually gets stored in any Java objects.  I haven't yet been able to see the actual structs in native memory.  I only see the memory address in the Eclipse debugger.

Thanks!

**Comment by neilharvey94044** [comment]
Just wanted to let you know I've figured out what was wrong.  

It's because JNR doesn't implement machine alignment in the structs.  It does a lot of work calculating alignment and looks like it is automatically handling it, but actually isn't.  The Struct class has a Padding object that is used to create padding where needed in structs.  But it means tediously going through and comparing the c structs with the JNR structs in a spreadsheet to determine where the padding belongs.  If there is a simpler way, I haven't found it.

I have implemented the solution for the DevParams in my local code, but still need to go through the exercise for the TunerParams.  When I test down deep in the RspDx params they are working, where they weren't before, so I think the offsets are accurate for that set of structs.  I'll be doing more testing until I'm confident I've got the offsets correct throughout.

This should allow me to easily implement other features...like antenna select, etc..

Thought you would like to know.  😊

**Comment by neilharvey94044** [comment]
For some reason the SDRPlay GainChange event seems to be getting filtered out, or not generated when running sdrtrunk.  Do you have any insight into why this might be happening?  Any help or guidance you can provide is much appreciated. I've been looking at it quite a while and see no reason for its omission.  In fact the PowerOverloadChange event is coming through just fine.  Also, the JNRMain.java produces GainChange events just fine, but not sdtrunk.  I've been comparing them but don't see any difference of consequence.  I want this event because I want to display the total gain calculated by the API, something very useful to know when making gain adjustments.

I should be wrapping up within a few days.  I've got all the JNR Structs aligned properly and everything appears to be working great.  The problem turned out to be JNR does a fine job of alignment inside Structs, but neglects to align at the end of a Struct when needed.  In the multi-layered SDRPlay structs, this causes misalignment on subsequent structs.

I don't have all the boundary constraints from SDRPlay, but as you said, it's a lot of useful function even without these being implemented.

Thanks!

**Comment by samphonic** [comment]
Glad you got it figured out!  I was puzzled by the alignment issues in your last post because _some_ of the fields are working fine so it's obviously doing some kind of alignment correctly.  But the end-struct issue explains why my less-deep implementation didn't run into those problems.  Nice work!

The only difference I can think of offhand between JNRMain and the sdrtrunk implementation is that sdrtrunk goes through the [StreamsReceiver interface](https://github.com/Sammy1Am/SDRplayJava/blob/master/src/main/java/io/github/sammy1am/sdrplay/StreamsReceiver.java), and gets called [here in the SDRplayDevice class.](https://github.com/Sammy1Am/SDRplayJava/blob/17347cdc78c7c128f993528e9b21787e8a72dcb0/src/main/java/io/github/sammy1am/sdrplay/SDRplayDevice.java#L76-L81).  I don't see anything in either of those places that should be making a distinction between the PowerOverloadChange and GainChange events though, so if one if making it through I would expect the other to as well.  Taking a look at those SDRplayDevice lines while sdrtrunk is running though would be where I'd start (in the event those are getting properly triggered but the event isn't making it out of there.

Can't wait until I get my antenna set up so I can try your changes (and test with the RSPduo I got after I'd already packed things away so I haven't tried at all yet).

**Comment by neilharvey94044** [comment]
Looks like you will be the one implementing the RSPduo and its multiple tuners.  Ouch, that appears to be quite a challenge.  I'm not planning to implement any device specific function for a device I don't own (RSP1A and RSPdx) and can test with, so that will leave the RSPduo to you.  😳

The debugger is useless when running the whole app, too many threads running concurrently, so l ast night I put some println's around the call to streamsReciever.receiveEvent in the SDRPlayDevice class's "call" method Override.  For the PowerOverLoadChange event both printlns are executed, for the GainChange event the first is executed and the second is not.  So the source code that indicates the call will be made is not aligning with the byte code.  Something strange is going on.  This is a call going from the SDRplayJava code to the sdrtrunk code, so that might mean something. 

I noticed you have some comments in the code about the event parameter data not using JNR, maybe this is doing something unexpected and different between event types?  Maybe use JNR to access native data and populate the managed objects?  Anyway, looks like a mystery, not intentional.  Next step is to try running outside of Eclipse with a release build and see if anything is different.  Then try/catches around this code.  Then if all that fails, I'll look at how these event objects are constructed.

I need to be quick though, I may end up going back to work on a contract shortly.  If I can't get it solved soon I will likely just push the code without it.

Btw, are you in the U.S. or elsewhere?  City? 

Thanks!

===========where the mystery code jump occurs=============

CallbackFnsT.EventCallback ecb = new CallbackFnsT.EventCallback() {
@Override
public void call(EventT eventId, TunerSelectT tuner, EventParamsT params, Pointer cbContext) {
System.out.println("SDRPlayDevice ---- Received Event Callback:" + eventId.name());

streamsReceiver.receiveEvent(eventId, tuner, new EventParameters(params));

System.out.println("SDRPlayDevice ---- After Invoke of streamReceiver.receiveEvent: " + eventId.name());
}
};

**Comment by neilharvey94044** [comment]
Found it.  Need to investigate, but this is a start.  Looks like a memory issue.  Strange that this exception isn't getting caught and reported somewhere up the stack.

SDRPlayDevice ---- Received Event Callback:GainChange
Exception caughtjava.lang.IllegalArgumentException: No known Enum mapping for value 50 of type io.github.sammy1am.sdrplay.jnr.CallbackFnsT$PowerOverloadCbEventIdT
SDRPlayDevice ---- Received Event Callback:GainChange

**Comment by samphonic** [comment]
Weird, it might be a memory thing, but it may also be something I overlooked in the Enums (which were a bit weird too).  I'm puzzled by the fact that it clearly knows it's a GainChange event, but then complains there's no value 50 for the `PowerOverloadCbEventIdT` type.

I'm looking forward to the multiple-tuner implementation actually.  It'll be a fun challenge to see if I can get sdrtrunk to actually treat the tuners as separate tuners (though adding a setting to toggle between the two modes will be a challenge with the way sdrtrunk currently configured tuners).

I'm up in the Pacific Northwest south of Seattle (did my stint at Amazon of course, but moved for my wife's work and I'm looking to get into some part-time work once we're more settled from the move).

**Comment by neilharvey94044** [comment]
I resolved it.  I think there were two things going on.  One, I think, was passing the JNR event parameters  down the call stack.  This may not have been a problem, but I changed it in any case to convert to managed objects as soon as the event is received.  You already had the code in place to do the conversion, but you were doing it later in the call stack.  I think the real problem was the code that converts to managed objects was not checking to see the type of event, so was initializing the managed object with values that were not valid for the enumerations.  SDRPlay passes a union, so only one set of values is valid depending upon the event type.  This was hidden because an uncaught exception was being thrown.

The one remaining problem is when I close the application a spurious overload event is passed by the API and being acted upon, but the API has already been unitialized, which results in an exception being thrown/caught/reported.    For some reason the isInitialized flag is not keeping the code from acknowledging the overload after the API is unitialized.   I think it's a threading/timing issue.  The thread in sdrtrunk that is managing SDRplayJava is not the same thread receiving overload events.  I'll keep looking at it, but at some point will just let this one go for another day since it's on termination.  The change I made above is probably what causes this...the events are sticking around longer because they are managed sooner.  The ideal solution is to figure out why they are generated in the first place, something is occurring in the API to cause an overload on shutdown.

We aren't far from each other.  I live about 15 miles south of San Francisco on the coast.   My wife and I considered moving to Seattle many years ago, it's beautiful country.  But we took a trip and discovered that lots of people in that area were depressed from lack of sunshine.  So we decided to stay in the SF bay area where we both grew up.

I'm going to try and have all of this pushed by the end of the weekend.

Thanks!

**Comment by neilharvey94044** [comment]
I have finished and submitted pull requests to SDRplayJava and sdrtrunk.

I finally decided to not generate exceptions when acknowledging overload conditions, just ignoring the error.  Error handling for this call is not really necessary anyway.

**Comment by samphonic** [comment]
> I think the real problem was the code that converts to managed objects was not checking to see the type of event, so was initializing the managed object with values that were not valid for the enumerations.

I spotted that in the changes and I'm not 100% sure what I was thinking there.  I guess assuming that the incoming native values would resolve to sane defaults if they were unused/uninitialized, but that wasn't a good assumption.

> a spurious overload event

Yeah, I didn't see that when I was testing, but I think you're right that the change you made may have affected when it gets handled to put it after the uninitialization.  Based on my understanding, the overload event is more of a nice-to-have thing to inform the user that their signal may be negatively impacted rather than something that needs to be treated as an error.  In that regard, I think discarding (or immediately catching and sending to a log statement) exceptions that happen while trying to acknowledge overload changes is a reasonable position to take.  Knowing why there's an extra overload event at the end would be nice, but _any time_ this happens I think swallowing the exception is the right way to go.

Thanks again for all your work on this, I'm sure SDRplay support for sdrtrunk is going to make a lot of people happy.

It is indeed beautiful up here, but it's also pretty much cloudy from October to May so if you like having sun regularly it's not ideal 😉.
